near-sdk/src/collections/vec/mod.rs

```diff
@@ -0,0 +1,258 @@
+//! A vector implemented on a trie. Unlike standard vector does not support insertion and removal
+//! of an element results in the last element being placed in the empty position.
+// TODO update these docs
+
+mod impls;
+mod iter;
+
+use self::iter::{Iter, IterMut};
+use crate::collections::append_slice;
+use crate::{env, CacheCell, CacheEntry, EntryState, IntoStorageKey};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::collections::{btree_map::Entry, BTreeMap};
+use std::ptr::NonNull;
+
+const ERR_INCONSISTENT_STATE: &[u8] = b"The collection is an inconsistent state. Did previous smart contract execution terminate unexpectedly?";
+const ERR_ELEMENT_DESERIALIZATION: &[u8] = b"Cannot deserialize element";
+const ERR_ELEMENT_SERIALIZATION: &[u8] = b"Cannot serialize element";
+const ERR_INDEX_OUT_OF_BOUNDS: &[u8] = b"Index out of bounds";
+
+fn expect_consistent_state<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| env::panic(ERR_INCONSISTENT_STATE))
+}
+
+/// An iterable implementation of vector that stores its content on the trie.
+/// Uses the following map: index -> element.
+///
+/// This implementation will cache all changes and loads and only updates values that are changed
+/// in storage after it's dropped through it's [`Drop`] implementation.
+///
+/// TODO examples
+#[derive(BorshSerialize, BorshDeserialize)]
+#[cfg_attr(not(feature = "expensive-debug"), derive(Debug))]
+pub struct Vector<T>
```

Comment:
Yeah, `Vector` seems like a bad name. It's different from Vec, but it's uncleare why. 

My understanding is that collections here are very different from std's collections, in that they are persistent (as in, backed by the try directly). So it's more like an ORM than a collections.

I suggest choosing some perfix here, and use it for all collections. A couple of options would be `Trie` or `Stored`: `StoredVec`, `TrieHashMap`.

Allternatively, *maybe* rather than using `collections` module, it should be part of the `store` module? So that the call-site looks like this:

```
struct MyContrct {
    data: store::Vec<String>
}

fn do_something() {
    store::write(..)
}
```

Reply:
Would it be helpful to try to design the whole API holistically holistically? That is, creating a docmuent which lists only signatures and bodies of all the collections?

Reply:
yeah for naming, I don't have an opinion, happy to change to whatever makes sense. I just favoured using the existing name to be as close to drag and drop from what it was (except for some APIs now closer to std).

As for documenting, absolutely will do, this first day was just exploratory to see what's possible before I narrowed down the API